### PR TITLE
Clean up the registration of recording and stimulation backends

### PR DESCRIPTION
**Pull request by jougs**
This is a rather unsorted collection of changes related to the registration of recording and stimulation backends in preparation of the module system refactoring.

### CODE REVIEW

File: nestkernel/io_manager_impl.h [hunk 1]
```diff
@@ -27,28 +27,35 @@
...
+    throw BackendAlreadyRegistered( name.toString() );
   }
+
+  StimulationBackendT* stimulation_backend = new StimulationBackendT();
+  stimulation_backend->pre_run_hook();
+  stimulation_backends_.insert( std::make_pair( name, stimulation_backend ) );
 }
-}
+
+} // namespace nest
```
Comment (med-ayssar) [inline nestkernel/io_manager_impl.h:59]: I would do the following:

```cpp
Class IOManager 
{
    template<class BackendT> std::map<Name, BackendT> get_backend(const BackendT& backend);
    template<class BackendT> void register_backend(const Name& name);

};

 template<> std::map<Name, StimulationBackend> IOManager::get_backend(const StimulationBackend& ){ return stimulation_backends_;}

 template<> std::map<Name, RecordingBackend> IOManager::get_backend(const RecordingBackend& ){ return recording_backends_;}
```
And then you have only one `register_backend` function that does the same for both recording and stimulation devices, depending on the type for the template class.

File: nestkernel/io_manager.cpp [hunk 2]
```diff
@@ -327,7 +336,7 @@ IOManager::enroll_recorder( const Name& backend_name, const RecordingDevice& dev
 }
 
 void
-nest::IOManager::enroll_stimulator( const Name& backend_name, StimulationDevice& device, const DictionaryDatum& params )
+nest::IOManager::enroll_stimulator( const Name backend_name, StimulationDevice& device, const DictionaryDatum& params )
 {
```
[the remark this answers is not in the document]
Comment (JanVogelsang) [reply nestkernel/io_manager.cpp:340]: Name is just an unsigned with some functions to interpret that int, so using a reference makes no sense here. It increased the size of the parameter and adds and indirection. As the Name is not modified in either of these functions, we can safely remove the reference.

File: nestkernel/io_manager.cpp [hunk 3]
```diff
@@ -60,8 +61,16 @@ namespace nest
 IOManager::IOManager()
   : overwrite_files_( false )
 {
-  register_recording_backends_();
-  register_stimulation_backends_();
+  register_recording_backend< RecordingBackendASCII >( "ascii" );
```
[the remark this answers is not in the document]
Comment (jougs) [reply nestkernel/io_manager.cpp:64]: Hmmm... We could also remove the function argument altogether and take it from the instance/class inside of the function. I will look into this.

**Review by JanVogelsang — APPROVED** [review]

File: nestkernel/exceptions.h [hunk 4]
```diff
@@ -1242,6 +1242,26 @@ class BackendNotPrepared : public KernelException
   const std::string backend_;
 };
 
+class BackendAlreadyRegistered : public KernelException
```
[the remark this answers is not in the document]
Comment (med-ayssar) [reply nestkernel/exceptions.h:1245]: An `exception` namespace under `nest` namespace would be optimal!

File: nestkernel/io_manager_impl.h (hunk 1 above)
Comment (jougs) [reply nestkernel/io_manager_impl.h:59]: I am also unhappy about the code duplication for the different backends, but would like to leave this for now.

My hope is that we can implement a generic `ComponentRegistry` later on, which would then support everything that can be dynamically registered (e.g. neuron and synapse models, RNG types, structural plasticity growth curves, connection rules, recording and stimulation backends).

File: nestkernel/io_manager.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jougs) [reply nestkernel/io_manager.cpp:64]: Ah, the backends don't actually have a `get_ name()` function and I don't think this beautification would warrant a change of the class interface.

File: nestkernel/exceptions.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (jougs) [reply nestkernel/exceptions.h:1245]: I have created #2641 to keep track of this in the context of PyNEST NG.

**Comment by JanVogelsang** [comment]
@jougs Something went wrong there, the lines are duplicated now.

File: nestkernel/io_manager_impl.h (hunk 1 above)
Comment (JanVogelsang) [reply nestkernel/io_manager_impl.h:59]: That's fine I think. As long as we add that to a kernel-team todo-list somewhere. We should start collecting all our ideas for the next hackathon following the SLI-purge one.

File: nestkernel/io_manager.cpp [hunk 5]
```diff
@@ -286,22 +295,23 @@ IOManager::cleanup()
...
 
 bool
-IOManager::is_valid_recording_backend( const Name& backend_name ) const
+IOManager::is_valid_recording_backend( const Name backend_name ) const
 {
-  std::map< Name, RecordingBackend* >::const_iterator backend;
+  auto backend = recording_backends_.find( backend_name );
+  return backend != recording_backends_.end();
   backend = recording_backends_.find( backend_name );
   return backend != recording_backends_.end();
```
Comment (jougs) [inline nestkernel/io_manager.cpp:303]:
```suggestion
```

File: nestkernel/io_manager.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jougs) [reply nestkernel/io_manager.cpp:64]: I have written down the idea in #2642.

File: nestkernel/io_manager_impl.h (hunk 1 above)
Comment (jougs) [reply nestkernel/io_manager_impl.h:59]: I have captured my ideas in #2642. Please feel free to comment there in case you have further ideas.

**Comment by jougs** [comment]
@JanVogelsang: I removed the duplicated lines. Also thanks for your other input.
@med-ayssar: I think I have addressed all your points. Can you please re-review? Thanks!

**Review by med-ayssar — APPROVED** [review]
Looks fine now!

**Pull request merged**